gateway/src/apicast/policy/3scale_batcher/reports_batcher.lua

```diff
@@ -0,0 +1,91 @@
+local keys_helper = require('apicast.policy.3scale_batcher.keys_helper')
+
+local setmetatable = setmetatable
+local ipairs = ipairs
+local insert = table.insert
+local resty_lock = require 'resty.lock'
+
+local _M = {}
+
+local mt = { __index = _M }
+
+local lock_timeout = 10
+local lock_options = { timeout = lock_timeout }
+
+-- Note: storage needs to implement shdict interface.
+function _M.new(storage, lock_shdict_name)
+  local self = setmetatable({}, mt)
+  self.storage = storage
+  self.lock_shdict_name = lock_shdict_name
+  return self
+end
+
+-- Note: When the lock cannot be acquired, the report will be lost.
+-- The timeout is high, that means that if it could not be acquired, there's
+-- probably a problem in the system.
+-- TODO: Find a solution for this.
+function _M:add(service_id, credentials, usage)
+  local deltas = usage.deltas
+
+  local lock, new_lock_err = resty_lock:new(self.lock_shdict_name, lock_options)
+  if not lock then
+    ngx.log(ngx.ERR, 'failed to create lock: ', new_lock_err)
+    return
+  end
+
+  local elapsed, lock_err = lock:lock(service_id)
+  if not elapsed then
+    ngx.log("failed to acquire the lock: ", lock_err)
+    return
+  end
+
+  for _, metric in ipairs(usage.metrics) do
+    local key = keys_helper.key_for_batched_report(service_id, credentials, metric)
+    self.storage:incr(key, deltas[metric], 0)
```

Comment:
This is able to silently delete data when it runs out of memory.

[`ngx.shared.DICT.incr`](https://github.com/openresty/lua-nginx-module#ngxshareddictincr) has no safe version and it mentions it deletes data

> Like the [add](https://github.com/openresty/lua-nginx-module#ngxshareddictadd) method, it also overrides the (least recently used) unexpired items in the store when running out of storage in the shared memory zone.

So the safe way to increment a number is to first allocate a key by [`ngx.shared.DICT.safe_add`](https://github.com/openresty/lua-nginx-module#ngxshareddictsafe_add) and then increment it.

And verify the `forcible` value `incr` returns and print huge warning that it deleted some other reports.